### DISCUSSION TITLE
blog/2022-04-30-static-builds/

### DISCUSSION THREAD

**Discussion by giscus** [crazychenz/vinnie.work#27] [vinnie.work Comments]
# blog/2022-04-30-static-builds/

Static builds, technically linking of binaries in such a way that there is no dynamic loading of external libraries to start the process. A more broad way that I prefer to think of static applications are ones that have no external userspace dependencies. One may argue that my definition of a static application could be a docker container. Not true! A docker container depends on docker (i.e. an external userspace dependency).

https://www.vinnie.work/blog/2022-04-30-static-builds/

**Comment by sfroberg** [comment]
Very interesting writing.
But no need to do qemu, even on x86_64 64-bit development host:

I have had some success previously with chrooted x86 Alpine setup.

Steps I took:
1. Goto https://alpinelinux.org/downloads/ and from there choose Mini Root Filesystem 
(https://dl-cdn.alpinelinux.org/alpine/v3.16/releases/x86/alpine-minirootfs-3.16.2-x86.tar.gz )

2. Extract & cd to minirootfs and chroot:

tar -xaf alpine-minirootfs-3.16.2-x86.tar.gz && cd alpine_x86
setarch linux32 chroot . /usr/bin/env -i HOME=/root /bin/sh

This makes sure that chroot environment is totally clean and not polluted by host environment variables (you can check by giving env command, should only list SHLVL, HOME, & PWD).

3. cd to root directory make sure all the normal development tools are installed and install whatever library you want statically (note alpine linux has not libraries statically linked but many):

cd
apk add gcc binutils m4 patch perl autoconf automake

For example, check if alpine linux has libpng as static version:
apk search libpng | grep static (should give libpng-static-1.6.37-r1)

Now most (not all, but most) stuff you compile inside that root is A) 32-bit so should work with both x86/x86_64
and statically compiled (uses musl libc)

If some application is difficult I do:
make CC="gcc -static"

99% of times it get's things solved

Enjoy

**Reply by crazychenz** [reply]
I completely agree that a chroot is another great way to contain a build environment!, especially when your x86 Linux distribution doesn't explicitly support multi-arch. Apologies if anything that follows is stuff you already are aware of...

`chroot` is dependent on the running kernel (i.e. its stuck using syscalls defined in Linux 2.6 and after). If you want to go back in time to Linux 2.4 or Linux 2.2, you may have to use virtualization (or dedicated hardware) so that the target libc's syscalls match the running kernel syscalls. The largest set of impacted syscalls between 2.2/2.4 and 2.6 (IMHO) is the network syscalls. And FWIW, I've seen quite a few embedded systems running Linux 2.4 to this day.

When suggesting chroot in contrast to qemu, I _assume_ you are responding to the comment: "With static qemu emulation, I'm aiming to portably containerize i386 systems." To be honest, I can't remember the exact idea that I was attempting to convey... :-P 

What I will say is that, building a static toolchain of binutils/gcc-2.95/glibc on a modern distribution is a non-trivial task. (See [crosstools](http://www.kegel.com/crosstool/) for more info.) I believe that most in the business would rather locate a very old distribution (circa 2000-2008) of Linux and hope there is enough support to build the required tools. Other groups have cached ancient OpenWRT or LFS builds that just work. Unfortunately, links in build systems like OpenWRT and LFS grow stale. I wrote a bit about this in [My Obsession With Static Builds](https://www.vinnie.work/blog/2022-05-01-building-to-the-version-generation/)

Lastly, I've not really been an alpine user. I'll keep your suggestion in mind. Thanks for the tip!

**Reply by sfroberg** [reply]
Hi.

Yes you right.
I completely forgot that kernel needs to be same if using that chroot trick. I was not sure if you only needed,mostly, portable Linux binaries in circa 2022 AD that would work on most 32-bit/64-bit x86/x86_64 system. Then musl libc + static linking is the way to go. 
Embedded systems are of course different story. And glibc is notoriously difficult, almost hostile I would say, to do static linking.

Speaking of old kernels and 32-bit, I vaguely remember that i386 support was dropped in 3.7 or 3.8 and now they going to, maybe, drop i486 support too(???)
https://www.phoronix.com/news/Intel-i486-Linux-Possible-Drop

I have been using Linux since 1998 and always valued the choice it offers. But this current trend of dropping support for 32-bit machines (I have several, including old Intel Atom Netbooks here) is troubling to me :(
I think am too old to learn FreeBSD ...


Very Best Regards:
S.F.

**Reply by sfroberg** [reply]
Also, quoting that phoronix article:
_"Not that I'm convinced most distros even do 32-bit development anyway these days."_
I can confirm that. I just yesterday tried Linux kernels 5.15, 5.19 and even the latest 6.0 series to compile on 32-bit Intel Atom arch with ZSTD compression enabled. None of them worked.

So it seems that 32-bit support is really left to drop and it's all just 64-bit in x86-land now, which is a shame, there still some nice machines (Pentium 2, 3, Celeron, Intel Atom) out there that would have some extra life left and not end to landfill ....

**Reply by crazychenz** [reply]
32 bit Intel is gone. ARM32, MIPS, and other 32 bit architectures are still trucking along!
